jobs/src/test/scala/dcos/metronome/SimulatedScheduler.scala

```diff
@@ -0,0 +1,82 @@
+package dcos.metronome
+
+import akka.actor.{ Cancellable, Scheduler }
+import java.util.concurrent.atomic.AtomicLong
+
+import dcos.metronome.utils.time.FixedClock
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration.FiniteDuration
+
+/**
+  * Simulates a scheduler using a FixedClock
+  *
+  * Does not use a separate thread for scheduling; however, scheduled tasks are executed according to their provided
+  * ExecutionContext. If using a same-thread execution context, then the task will be executed by the thread that
+  * advances the clock.
+  *
+  * If scheduler checks if tasks need to be run each time the clock is advanced. If a repeating task is scheduled to
+  * happen every 10 seconds, and the clock is advanced 1 minute, then the task will be fired off once and the next run
+  * be scheduled for 10 seconds later. This mirrors the behavior of the Akka scheduler when it cannot fire tasks off as
+  * quickly as it is asked to.
+  */
+class SimulatedScheduler(clock: FixedClock) extends Scheduler {
+  override def maxFrequency = 0.0
+  private[this] val nextId = new AtomicLong
+  private[this] val scheduledTasks = scala.collection.mutable.Map.empty[Long, ScheduledTask]
+  private case class ScheduledTask(action: () => Unit, var time: Long)
+  private class ScheduledTaskCancellable(id: Long) extends Cancellable {
+    override def cancel() = {
+      doCancel(id)
+      true
+    }
+    override def isCancelled = scheduledTasks.contains(id)
+  }
+
+  clock.onChange { () => poll() }
+
+  private[this] def doCancel(id: Long) = synchronized { scheduledTasks -= id }
+  private[this] def poll(): Unit = synchronized {
+    val now = clock.now().getMillis
+    scheduledTasks.values.foreach { task =>
+      if (task.time <= now) task.action()
```

Comment:
sigh; I wrote this code originally. And there is a deadlock potential here if the Executor provided is a same-thread executor. I guess keep for now and I'll fix it in Marathon / copy over in the future.

Reply:
Simulated scheduler deadlock potential fixed here: https://github.com/mesosphere/marathon/pull/5879

Reply:
@timcharper I looked into resolving this and it does seem as simple as the PR you provided as reference.   I've created a JIRA to capture the tech debt:  https://jira.mesosphere.com/browse/METRONOME-209